.emacs.d/config.org

```diff
@@ -55,62 +55,404 @@
+
+  #+BEGIN_SRC emacs-lisp
+  (use-package flycheck
+    :hook (after-init . global-flycheck-mode)
+    :custom
+    (flycheck-check-syntax-automatically '(save mode-enabled)))
+  #+END_SRC
+
+** Documentation
+
+  When the point is overlapping with a function call, eldoc shows the function arguments in the echo area.
+
+  #+BEGIN_SRC emacs-lisp
+  (use-package eldoc-mode
+    :hook (after-init . global-eldoc-mode))
+  #+END_SRC
+
+* Language and debug servers
+
+  [[https://microsoft.github.io/language-server-protocol][LSP]] is a protocol used to implement IDE-like features in an editor-agnostic manner. This package will give us a uniform way to add features such as semantic highlighting, auto-completion, jump-to-definition, and so on.
+
+  #+BEGIN_SRC emacs-lisp
+  (use-package lsp-mode
+    :commands lsp
+    :custom
+    (lsp-auto-guess-root t)
+    (lsp-log-io t)
+    (lsp-prefer-flymake nil))
+  #+END_SRC
+
+  This package enhances the user interface, providing pop-up windows with documentation and visual indicators.
+
+  #+BEGIN_SRC emacs-lisp
+  (use-package lsp-ui
+    :hook (lsp-mode . lsp-ui-mode)
+    :custom
+    (lsp-ui-doc-delay 3.0)
+    (lsp-ui-doc-include-signature t)
+    (lsp-ui-flycheck-enable t)
+    (lsp-ui-peek-always-show t))
+  #+END_SRC
+
+  Analogous to LSP is the [[https://microsoft.github.io/debug-adapter-protocol/][debug adapter protocol]]. We use [[https://github.com/emacs-lsp/lsp-treemacs][treemacs]] for IDE-like display of errors.
+
+  #+BEGIN_SRC emacs-lisp
+  (use-package dap-mode
+    :after lsp-mode
+    :config
+    (dap-mode t)
+    (dap-ui-mode t))
+
+  (use-package treemacs
+    :bind (:map global-map
+                ("C-x t t" . treemacs)
+                ("C-x t 1" . treemacs-select-window))
+    :custom
+    (treemacs-resize-icons 15))
+
+  (use-package lsp-treemacs
+    :bind (:map java-mode-map
+                ("C-x e l" . lsp-treemacs-errors-list)
+                ("C-x s l" . lsp-treemacs-symbols))
+    :init
+    (lsp-treemacs-sync-mode 1))
+  #+END_SRC
+
+* Code completion
+
+  Company is the primary package that is used for code completion, it follows a frontend/backend system. The package =company= is the frontend, it will query a certain backend based on what code you are editing, such as one provided by an active language server.
+
+  #+BEGIN_SRC emacs-lisp
+  (use-package company
+    :hook (after-init . global-company-mode)
+    :custom
+    (company-tooltip-align-annotations t))
+
+  (use-package company-lsp
+    :after (company lsp-mode)
+    :custom
+    (company-lsp-cache-candidates t))
+  #+END_SRC
+
+* Languages
+
+** HTML / CSS / HTML Templates
+
+- [[https://docs.emmet.io/cheatsheet-a5.pdf][Emmet cheatsheet]]
+- [[https://www.youtube.com/watch?v=V8vizNQKtx0][Learn Emmet in 15 Minutes (YouTube)]]
+
+  #+BEGIN_SRC emacs-lisp
+  (use-package emmet-mode
+    :hook ((css-mode php-mode sgml-mode rjsx-mode web-mode) . emmet-mode))
+  #+END_SRC
+
+** JavaScript / TypeScript
```

Comment:
Disable some js/ ts flycheck checks which will automatically be handled by prettier/ linter. 
```lisp
(setq js2-missing-semi-one-line-override t)
(setq js2-strict-missing-semi-warning nil)
```